astropy/units/equivalencies.py

```diff
@@ -371,40 +380,21 @@ def doppler_radio(rest):
     <Quantity -31.209092088877583 km / s>
     """
     assert_is_spectral_unit(rest)
-
-    ckms = _si.c.to_value("km/s")
-
-    def to_vel_freq(x):
-        restfreq = rest.to_value(si.Hz, equivalencies=spectral())
-        return (restfreq - x) / (restfreq) * ckms
-
-    def from_vel_freq(x):
-        restfreq = rest.to_value(si.Hz, equivalencies=spectral())
-        voverc = x / ckms
-        return restfreq * (1 - voverc)
-
-    def to_vel_wav(x):
-        restwav = rest.to_value(si.AA, spectral())
-        return (x - restwav) / (x) * ckms
-
-    def from_vel_wav(x):
-        restwav = rest.to_value(si.AA, spectral())
-        return restwav * ckms / (ckms - x)
-
-    def to_vel_en(x):
-        resten = rest.to_value(misc.eV, equivalencies=spectral())
-        return (resten - x) / (resten) * ckms
-
-    def from_vel_en(x):
-        resten = rest.to_value(misc.eV, equivalencies=spectral())
-        voverc = x / ckms
-        return resten * (1 - voverc)
-
+    spectral_eq = functools.partial(rest.to_value, equivalencies=spectral())
+    to_funcs = {
+        si.Hz: lambda x: (1 - x / spectral_eq(si.Hz)) * ckms,
+        misc.eV: lambda x: (1 - x / spectral_eq(misc.eV)) * ckms,
+        si.AA: lambda x: (1 - spectral_eq(si.AA) / x) * ckms,
+    }
+    from_funcs = {
+        si.Hz: lambda x: spectral_eq(si.Hz) * (1 - x / ckms),
+        misc.eV: lambda x: spectral_eq(misc.eV) * (1 - x / ckms),
+        si.AA: lambda x: spectral_eq(si.AA) / (1 - x / ckms),
+    }
```

Comment:
Absolutely not required, but you could consider combining `to_funcs` and `from_fI had to think about this: it is conventient to define the partial funcs` definitions into
```
funcs = {si.Hz: (<to_func>, <from_func>), ...}
```
and then use `*func` in the list creation, avoiding `to_func, from_funcs[unit]` (I considered suggesting iterating also over `from_funcs.values()` but dislike that that counts on order being the same).

Of course, at that point, you could also consider just returning the whole slew in one go...

Reply:
I found it better to first group the functions based on the direction of the conversion instead of the unit because this way similar functions are next to each other.